lib/user_space/user_x86_64_atomic.h

```diff
@@ -167,11 +254,43 @@ static inline bool m0_atomic64_cas(int64_t * loc, int64_t oldval, int64_t newval
 		     : "r" (newval), "0" (oldval)
 		     : "memory");
 	return val == oldval;
+#elif defined CONFIG_AARCH64 /* aarch64 */
+/*	unsigned long	tmp;
+	int64_t	 old=0;
+
+	M0_CASSERT(8 == sizeof oldval);
+
+	asm volatile("// atomic64_cas_return \n"		\
+		     "  prfm    pstl1strm, %[v]\n"		\
+		     "1:ldxr    %[old], %[v]\n"  		\
+		     "  eor     %[tmp], %[oldval], %[old]\n"	\
+		     "  cbnz    %[tmp], 2f\n"			\
+		     "  stxr    %w[tmp], %[newval], %[v]\n"  	\
+		     "  cbnz    %w[tmp], 1b\n"			\
+		     "  \n" 					\
+		     "2:"					\
+		     : [tmp] "=&r" (tmp), [oldval] "=&r" (oldval), \
+		       [v] "+Q" (*(unsigned long *)loc) 	\
+		     : [old] "Lr" (old), [newval] "r" (newval)  \
+		     :);
+	return old == oldval;// need to be reviewed
+#endif
+	return 0;
+*/
+
+	M0_CASSERT(8 == sizeof oldval);
+	return __atomic_compare_exchange_n(loc, &oldval, newval, 0, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+#endif
+	return 0;
 }
 
 static inline void m0_mb(void)
 {
+#ifdef CONFIG_X86_64
 	asm volatile("mfence":::"memory");
+#elif defined CONFIG_AARCH64 /* aarch64 */
+	asm volatile("dsb sy":::"memory");
+#endif
```

Comment:
In case of some other architecture (not x86 or aarch64) - this code would do nothing, and it will compile silently without any complaint. And it will be very hard to find out then why something is not working.

So do it like this:

```
#if defined CONFIG_AARCH64
	asm volatile("dsb sy":::"memory");
#else /* CONFIG_X86_64 */
	asm volatile("mfence":::"memory");
#endif

This way it will give a compilation error at least when some new platform is introduced.

Reply:
The code  that was there before  was in similar  structure with respect to x86  macro variable , but  for the sake  of explicitly checking for platform, we have  introduced this current structure. if new platform is going to be introduced, the  build chain will not work  unless configure.ac platform is configured. The  current code is however Ok, we can stick to this structure

Reply:
I think I have explained why this code is dangerous and bugs-prone. Which is still holds even with the configure.ac, because even when some new platform is introduced there - you still have to find and update the code in many places. And this place may be easily missed because there won't be any compilation problem on it in its current form.

Please fix it as suggested.

Reply:
This is also true for all places with the "if ... elif ... fi".
 If a new platform is introduced in configure.ac, but not updated here, the compile will still succeed, but this is a bug.

Reply:
it is mandatory to keep the configure.ac updated as per the platforms supported

Reply:
Yes, but after configure.ac is updated the code will compile just fine - and that's the problem (because the code must be updated as well, and it should not compile if it's not updated).

Looks like you've fixed this, so I'm resolving it.